File: c++/src/celf.cpp
#include "im/celf.hpp"

#include <algorithm>
#include <chrono>
#include <iostream>
#include <queue>
#include <stdexcept>
#include <vector>

namespace im {
namespace {

struct CELFNodeCompare {
    bool operator()(const CELFNode& left, const CELFNode& right) const {
        if (left.marginal_gain != right.marginal_gain) {
            return left.marginal_gain < right.marginal_gain;
        }
        if (left.iteration != right.iteration) {
            return left.iteration > right.iteration;
        }
        return left.node > right.node;
    }
};

double elapsed_seconds(const std::chrono::steady_clock::time_point& start_time) {
    const auto end_time = std::chrono::steady_clock::now();
    const auto duration = std::chrono::duration_cast<std::chrono::duration<double>>(end_time - start_time);
    return duration.count();
}

}  // namespace

CelfResult celf(
    const Graph& graph,
    const CelfParams& params,
    const std::vector<ReachabilityGraph>* pre_sampled_subgraphs) {
    if (pre_sampled_subgraphs == nullptr && params.num_subgraphs <= 0) {
        throw std::invalid_argument("num_subgraphs must be a positive integer");
    }
    if (pre_sampled_subgraphs != nullptr && pre_sampled_subgraphs->empty()) {
        throw std::invalid_argument("pre_sampled_subgraphs must be non-empty when provided");
    }

    CelfResult result;
    const int total_nodes = node_count(graph);
    if (total_nodes == 0 || params.k <= 0) {
        return result;
    }

    const int effective_k = std::min(params.k, total_nodes);
    std::vector<ReachabilityGraph> local_subgraphs;
    const std::vector<ReachabilityGraph>* sampled_subgraphs = pre_sampled_subgraphs;

    const std::optional<int> resolved_subgraph_seed =
        params.subgraph_seed.has_value() ? params.subgraph_seed : params.seed;

    if (sampled_subgraphs == nullptr) {
        const auto t_sample_start = std::chrono::steady_clock::now();
        local_subgraphs = generate_live_edge_subgraphs(
            graph,
            params.p,
            params.num_subgraphs,
            resolved_subgraph_seed);
        result.timing.phase_sample_subgraphs = elapsed_seconds(t_sample_start);
        result.timing.subgraph_source = "generated";
        sampled_subgraphs = &local_subgraphs;

        if (params.verbose) {
            std::cout << "[CELF] Phase 0 done: sample "
                      << params.num_subgraphs
                      << " subgraphs in "
                      << result.timing.phase_sample_subgraphs
                      << "s"
                      << '\n';
        }
    } else {
        result.timing.phase_sample_subgraphs = 0.0;
        result.timing.subgraph_source = "external";
        if (params.verbose) {
            std::cout << "[CELF] Phase 0 skipped: reuse "
                      << sampled_subgraphs->size()
                      << " pre-sampled subgraphs"
                      << '\n';
        }
    }

    result.timing.subgraph_count = static_cast<int>(sampled_subgraphs->size());
    if (sampled_subgraphs->empty()) {
        return result;
    }

    const auto t_init_start = std::chrono::steady_clock::now();
    std::priority_queue<CELFNode, std::vector<CELFNode>, CELFNodeCompare> queue;

    for (int u = 0; u < total_nodes; ++u) {
        std::set<int> singleton_seed_set;
        singleton_seed_set.insert(u);
        const double marginal_gain = average_spread_on_live_edge_subgraphs(*sampled_subgraphs, singleton_seed_set);
        queue.push(CELFNode{u, marginal_gain, 0});
    }

    result.timing.phase_init_queue = elapsed_seconds(t_init_start);

    if (params.verbose) {
        std::cout << "[CELF] Phase 1 done: build queue in "
                  << result.timing.phase_init_queue
                  << "s"
                  << '\n';
    }

    double current_spread = 0.0;
    const auto t_loop_start = std::chrono::steady_clock::now();

    while (static_cast<int>(result.selected.size()) < effective_k && !queue.empty()) {
        const int current_iteration = static_cast<int>(result.selected.size());

        CELFNode best_node = queue.top();
        queue.pop();
        const int u = best_node.node;

        if (best_node.iteration == current_iteration) {
            result.selected.insert(u);
            current_spread += best_node.marginal_gain;

            if (params.verbose) {
                std::cout << "[CELF] k="
                          << result.selected.size()
                          << "/"
                          << effective_k
                          << " selected node="
                          << u
                          << ", marginal_gain="
                          << best_node.marginal_gain
                          << '\n';
            }
        } else {
            const auto t_recompute_start = std::chrono::steady_clock::now();
            std::set<int> candidate_seed_set = result.selected;
            candidate_seed_set.insert(u);
            const double new_spread = average_spread_on_live_edge_subgraphs(*sampled_subgraphs, candidate_seed_set);
            result.timing.phase_recompute_only += elapsed_seconds(t_recompute_start);

            const double new_marginal_gain = std::max(0.0, new_spread - current_spread);
            queue.push(CELFNode{u, new_marginal_gain, current_iteration});
            result.timing.stale_recompute_count += 1;
        }
    }

    result.timing.phase_k_loop = elapsed_seconds(t_loop_start);

    if (params.verbose) {
        std::cout << "[CELF] Phase 2 done: k-loop in "
                  << result.timing.phase_k_loop
                  << "s"
                  << '\n';
        std::cout << "[CELF] Recompute detail: count="
                  << result.timing.stale_recompute_count
                  << ", time="
                  << result.timing.phase_recompute_only
                  << "s"
                  << '\n';
    }

    return result;
}

}  // namespace im

File: src/algorithms/celf.py
from __future__ import annotations

import heapq
import time
from dataclasses import dataclass
from typing import Dict, List, Set, Tuple

from src.common.shared import (
    average_spread_on_live_edge_subgraphs,
    generate_live_edge_subgraphs,
)


@dataclass
class CELFNode:
    node: int
    marginal_gain: float
    iteration: int

    def __lt__(self, other: "CELFNode") -> bool:
        if self.marginal_gain != other.marginal_gain:
            return self.marginal_gain > other.marginal_gain
        if self.iteration != other.iteration:
            return self.iteration < other.iteration
        return self.node < other.node


def celf(
    graph: Dict[int, List[int]],
    vertices: Set[int],
    k: int,
    p: float = 0.3,
    num_subgraphs: int = 50,
    seed: int | None = None,
    subgraph_seed: int | None = None,
    verbose: bool = True,
    pre_sampled_subgraphs: List[Dict[int, int]] | None = None,
) -> Tuple[Set[int], dict]:
    """Run CELF using pre-sampled live-edge subgraphs."""
    if pre_sampled_subgraphs is None and num_subgraphs <= 0:
        raise ValueError("num_subgraphs must be a positive integer")
    if pre_sampled_subgraphs is not None and not pre_sampled_subgraphs:
        raise ValueError("pre_sampled_subgraphs must be non-empty when provided")

    q: List[CELFNode] = []
    selected: Set[int] = set()
    current_spread = 0.0

    timing = {
        "phase_sample_subgraphs": 0.0,
        "phase_init_queue": 0.0,
        "phase_k_loop": 0.0,
        "phase_recompute_only": 0.0,
        "stale_recompute_count": 0,
        "subgraph_source": "generated",
        "subgraph_count": 0,
    }

    resolved_subgraph_seed = seed if subgraph_seed is None else subgraph_seed

    if pre_sampled_subgraphs is None:
        t_sample_start = time.perf_counter()
        sampled_subgraphs = generate_live_edge_subgraphs(
            graph,
            p=p,
            k=num_subgraphs,
            seed=resolved_subgraph_seed,
        )
        timing["phase_sample_subgraphs"] = time.perf_counter() - t_sample_start
        timing["subgraph_source"] = "generated"

        if verbose:
            print(
                "[CELF] Phase 0 done: sample "
                f"{num_subgraphs} subgraphs in {timing['phase_sample_subgraphs']:.4f}s"
            )
    else:
        sampled_subgraphs = pre_sampled_subgraphs
        timing["phase_sample_subgraphs"] = 0.0
        timing["subgraph_source"] = "external"
        if verbose:
            print(
                "[CELF] Phase 0 skipped: reuse "
                f"{len(sampled_subgraphs)} pre-sampled subgraphs"
            )

    timing["subgraph_count"] = len(sampled_subgraphs)

    if not sampled_subgraphs:
        return set(), timing

    t_init_start = time.perf_counter()
    for u in sorted(vertices):
        marginal_gain = average_spread_on_live_edge_subgraphs(sampled_subgraphs, {u})
        q.append(CELFNode(node=u, marginal_gain=marginal_gain, iteration=0))
    heapq.heapify(q)
    timing["phase_init_queue"] = time.perf_counter() - t_init_start

    if verbose:
        print(f"[CELF] Phase 1 done: build queue in {timing['phase_init_queue']:.4f}s")

    t_loop_start = time.perf_counter()
    announced_k = 0

    while len(selected) < k and q:
        current_iteration = len(selected)
        current_k = current_iteration + 1

        if verbose and current_k != announced_k:
            print(f"[CELF] Running k={current_k}/{k}...")
            announced_k = current_k

        best_node = heapq.heappop(q)
        u = best_node.node

        if best_node.iteration == current_iteration:
            selected.add(u)
            current_spread += best_node.marginal_gain
            if verbose:
                print(
                    f"[CELF] k={len(selected)}/{k} selected node={u}, "
                    f"marginal_gain={best_node.marginal_gain:.4f}"
                )
        else:
            t_recompute_start = time.perf_counter()
            new_spread = average_spread_on_live_edge_subgraphs(
                sampled_subgraphs,
                selected | {u},
            )
            timing["phase_recompute_only"] += time.perf_counter() - t_recompute_start

            new_marginal_gain = max(0.0, new_spread - current_spread)
            heapq.heappush(
                q,
                CELFNode(node=u, marginal_gain=new_marginal_gain, iteration=current_iteration),
            )
            timing["stale_recompute_count"] += 1

    timing["phase_k_loop"] = time.perf_counter() - t_loop_start

    if verbose:
        print(f"[CELF] Phase 2 done: k-loop in {timing['phase_k_loop']:.4f}s")
        print(
            "[CELF] Recompute detail: "
            f"count={timing['stale_recompute_count']}, "
            f"time={timing['phase_recompute_only']:.4f}s"
        )

    return selected, timing
